### ISSUE TITLE
thunderbird crash on start with v1.29.1/v1.29.2

### ISSUE THREAD

**Issue by rotanid** [vanowm/MasterPasswordPlus#136]
after updating masterpasswordplus to v1.29.1, Thunderbird doesn't start anymore.
OS: Arch Linux
Thunderbird 52.5.0-1

it crashes with a coredump:
https://gist.github.com/rotanid/8fa4211623cd4158a50c02e38be40d18

as soon as i disable masterpasswordplus (or downgrade it to v1.29) the issue is gone. (disabling/downgrading is done via safe-mode)

@vanowm tried to fix it with v.1.29.2 but the issue persists. (see #135)
i also disabled all other Add-Ons, but that didn't help either.

if a "strace" output would help, here's an excerpt. i don't want to put the whole thing online because there may be personal data in it.
https://gist.github.com/rotanid/55e6246188b96e682cabbcf511b1bdc4

**Comment by vanowm** [comment]
Can you do the following tests:

In about:config change `extensions.masterPasswordPlus.version` to 1.29.2 and then update MP+ to v1.29.2
If still crashing, try downgrade to v1.29 and reset `extensions.masterPasswordPlus.version` to empty string, if it's now crashing with 1.29 then the problem is related to updating the script and not necesarely related to the actual changes between v1.29 and 1.29.2
https://github.com/vanowm/MasterPasswordPlus/compare/a9134d26d169a6b542c6a42768d22555a00db5e4...008f1a6fec30970bfa491f6a8ce1cc28764a21ef


It could be helpful if you provide data from MP+ Options -> Help -> Troubleshooting info

**Comment by rotanid** [comment]
@vanowm there is no such setting `extensions.masterPasswordPlus.version`
if i create it manually and set it like you suggested, the problem doesn't go away.

also, the "troubleshooting info" text field is completely empty (with v1.29 of course, cant start with v1.29.2)

**Comment by vanowm** [comment]
Is there any settings with `extensions.masterPasswordPlus.` ?
If not, then there must be some errors in the console

**Comment by rotanid** [comment]
there are settings:
![bildschirmfoto_2017-12-10_02-34-14](https://user-images.githubusercontent.com/654617/33801168-a907cfb6-dd52-11e7-9dd0-7353588e3cb3.png)

**Comment by vanowm** [comment]
I assume this is with v1.29?
This shows that it's fails create default settings, due to [Bug 1414398](https://bugzilla.mozilla.org/show_bug.cgi?id=1414398), which strange considering that on Windows system TB 52.5.0 does not have this issue, it only starts with TB 58.0a1

Since I'm unable install Arch Linux in VMWare, only you can help find the cause of this problem...

If you would, install 1.29 and replace **masterpasswordtimeoutplus@vano\chrome\modules\masterpasswordplusCore.jsm** from [v1.29.2](https://github.com/vanowm/MasterPasswordPlus/blob/008f1a6fec30970bfa491f6a8ce1cc28764a21ef/src/chrome/modules/masterpasswordplusCore.jsm)

If it still crash TB, then try remove whole try{}catch{} block from that file line [1734](https://github.com/vanowm/MasterPasswordPlus/blob/008f1a6fec30970bfa491f6a8ce1cc28764a21ef/src/chrome/modules/masterpasswordplusCore.jsm#L1734) through [1777](https://github.com/vanowm/MasterPasswordPlus/blob/008f1a6fec30970bfa491f6a8ce1cc28764a21ef/src/chrome/modules/masterpasswordplusCore.jsm#L1777)

Also, you'll probably need delete `startupCache` folder from profile directory after each change you do to the extension.

**Comment by rotanid** [comment]
yes, this was with v1.29 of course, as TB doesn't start with newer ones.

i replaced the jsm file.
there was no crash, but the master password prompt didn't stop, no matter if i entered the correct or wrong passwords, it appeared again over and over.

then, i removed the try-catch-block you suggested. now TB starts after entering the password.

note though, that there is no folder named `startupCache` at all in the directory `.thunderbird`

**Comment by reaganch** [comment]
FYI, I just experienced the crash on Thunderbird startup with the addon enabled too. Disabling the addon from safe mode solves the problem.

**Comment by vanowm** [comment]
Ok, in v1.29.3b1 the changes made in 1.29.1 is now centralized, so it should be easier to track. Please try this version, if it still crashes, then for starters, comment out (or delete) this line:
https://github.com/vanowm/MasterPasswordPlus/blob/a04453715168492ba6cacd4b4c96ab8f3348fff2/src/chrome/modules/masterpasswordplusCore.jsm#L1820)
If you manage make it not crash, see what's in the browser console.

**Comment by rotanid** [comment]
@vanowm i'm sorry, but it still crashes with v1.29.3b1, even after commenting out line 1820 :-(

**Comment by vanowm** [comment]
From previous test, we determined the crash comes from that try{}catch{} section
So let's try determine what is in that section causing it.
comment out everything inside the catch(e){}
starting from https://github.com/vanowm/MasterPasswordPlus/blob/a04453715168492ba6cacd4b4c96ab8f3348fff2/src/chrome/modules/masterpasswordplusCore.jsm#L1799
through https://github.com/vanowm/MasterPasswordPlus/blob/a04453715168492ba6cacd4b4c96ab8f3348fff2/src/chrome/modules/masterpasswordplusCore.jsm#L1820

If still crashing, revert all changes and comment out just
https://github.com/vanowm/MasterPasswordPlus/blob/a04453715168492ba6cacd4b4c96ab8f3348fff2/src/chrome/modules/masterpasswordplusCore.jsm#L1795

If it **stops** crashing, then move that last line above the try{} section, maybe it will show something in the console...

P.S.
I've managed launch Arch Linux in VMWare (no help to it's installation guides), but it's just command line mode, I have no idea where to go from there.

**Comment by rotanid** [comment]
so i tried like you suggested, then i reverted and tried again, but now it worked fine, the troubleshooting info is now shown, also the "close" button from #137 does work.
i don't have any idea how this happened.
the only difference i can spot is: there are many more options now in "about:config" beginning with "extensions.masterpasswordplus" than were seen in the screenshot i posted before.
should i test anything again or should we close this? it's up to you.

**Comment by reaganch** [comment]
Sorry for the empty comment up there. I'm running Thunderbird 52.5.0 on Linux Mint 18.3. MasterPassword+ is at version 1.29.2 and currently disabled. No idea how to build the addon from source, so I think I'll just leave it disabled until an update fixes it I guess :-)

**Comment by rotanid** [comment]
@reaganch you don't have to build from source, the github project here contains builds for v1.29.3b1 too - i also didn't build anything.

**Comment by vanowm** [comment]
The only thing I could think of is there is some kind of cache going on. On Windows system there is a `startupCache` folder inside profile, I had to restrict it to readonly otherwise I coudln't properly test any changes I made to the addon.
Perhaps in the future it would be safer delete addon, restart the browser, install new version of the addon.

@reaganch there are .xpi packages of the addon:
https://github.com/vanowm/MasterPasswordPlus/tree/master/download
but you'll need enable `xpinstall.signatures.required` in about:config
Also, before you install anything from here, try uninstall MP+ and install it again from 
https://addons.mozilla.org/firefox/addon/master-password/

P.S.
which version are you running?
https://linuxmint.com/download.php

**Comment by reaganch** [comment]
Thanks a lot for the instructions! Uninstalled and reinstalled v1.29.2 of the addon, but that did not resolve the crashes. Toggling the `xpinstall.signatures.required` flag, uninstalling, and installing v1.29.3b1 of the addon still resulted in crashes. Am running Linux Mint 18.3 Cinnamon 64-bit.

**Comment by vanowm** [comment]
The TB version that comes with Mint Linux is v52.4.0 64bit
Installed MP+ 1.29.2 and it worked just fine...
Perhaps it's related to some MP+ settings. So, while MP+ is disabled, in `about:config` search for: `extensions.masterpasswordplus.`
If any of the settings shown and bold, post a screenshot of that window. (bold means modified from default).
From there you can also reset these settings via right click on each line and see if that helps at all..

**Comment by reaganch** [comment]
Yup, that worked :-)
I reset all the settings starting with `extensions.masterpasswordplus.` that were bolded, enabled the plugin, restarted Thunderbird and all is working fine.
I changed the addon settings to the way I like them, restarted again, and it's still okay.
Thanks so much for your help with this. Really appreciate it!

**Comment by vanowm** [comment]
Well...the problem is...now we don't know what settings caused it, unless you created a screenshot ;)

**Comment by reaganch** [comment]
Ah dammit! Didn't take a screenshot :-( So sorry. How hard can it be to follow instructions, right?

**Comment by etam** [comment]
I had the same problem. Thanks to your discussion I checked my config and compared to the defaults. 

I had this in config:
```
user_pref("extensions.masterPasswordPlus.lockhotkey", false);
user_pref("extensions.masterPasswordPlus.locklogouthotkey", false);
user_pref("extensions.masterPasswordPlus.lockwinhotkey", false);
user_pref("extensions.masterPasswordPlus.logouthotkey", false);
```

In defaults there is:
```
pref("extensions.masterPasswordPlus.lockhotkey", "ALT K");
pref("extensions.masterPasswordPlus.lockhotkeyenabled", 1);
pref("extensions.masterPasswordPlus.locklogouthotkey", "ALT SHIFT L");
pref("extensions.masterPasswordPlus.locklogouthotkeyenabled", 1);
pref("extensions.masterPasswordPlus.lockwinhotkey", "ACCEL ALT K");
pref("extensions.masterPasswordPlus.lockwinhotkeyenabled", 1);
pref("extensions.masterPasswordPlus.logouthotkey", "ALT L");
pref("extensions.masterPasswordPlus.logouthotkeyenabled", 1);
```

The type of all *hotkey prefs was changed from bool to string. I deleted those prefs and it's not crashing anymore.

**Comment by vanowm** [comment]
@etam thank you for the info! are you also using TB on Linux?

Weird part is I don't recall when these settings ever been boolean type, how did they end up as such? Perhaps when attempting save options while the extension was broken due to [Bug 1414398](https://bugzilla.mozilla.org/show_bug.cgi?id=1414398)?

I still can't reproduce it on fresh Linux Mint (or Windows)

@etam now that's working for you, can you show your info from MP+ options -> Help -> Troubleshooting

**Comment by etam** [comment]
> are you also using TB on Linux?

Yes

> can you show your info from MP+ options -> Help -> Troubleshooting

```
{
  "Addon": "Master Password+ v1.29.2",
  "Program": "Thunderbird v52.5.0 (en-US)",
  "OS": "Linux (x86_64-gcc3)",
  "Preferences": {
    "command": 0,
    "commandloggedin": true,
    "contextmenu": false,
    "debug": 1,
    "failedattempts": 5,
    "failedattemptstime": 10,
    "forceprompt": "[{\"enabled\":\"1\",\"name\":\"PassHash\",\"url\":\"chrome://passhash/content/passhash-dialog.xul\",\"param\":\"\"},{\"enabled\":\"1\",\"name\":\"FireFTP\",\"url\":\"chrome://fireftp/content/fireftp.xul\",\"param\":\"\"}]",
    "hidenewmailalert": true,
    "hidenewmailballoon": true,
    "identify": "",
    "idle": 3,
    "lockbgimage": true,
    "locked": false,
    "lockhidetitle": true,
    "lockhotkey": "ALT K",
    "lockhotkeyenabled": 0,
    "lockinactivity": true,
    "lockincorrect": 3,
    "locklogouthotkey": "ALT SHIFT L",
    "locklogouthotkeyenabled": 0,
    "lockminimize": false,
    "lockminimizeblur": true,
    "lockonminimize": 0,
    "lockonsleep": false,
    "lockrestore": true,
    "locktimeout": 10,
    "locktimer": false,
    "locktransparent": true,
    "lockwinhotkey": "ACCEL ALT K",
    "lockwinhotkeyenabled": 0,
    "logouthotkey": "ALT L",
    "logouthotkeyenabled": 0,
    "logoutinactivity": true,
    "logoutonminimize": false,
    "logoutonsleep": false,
    "logouttimeout": 10,
    "logouttimer": false,
    "minimizenoflicker": true,
    "nonlatinwarning": 2,
    "noworkaround": "",
    "persistnolock": true,
    "protect": false,
    "showchangeslog": 1,
    "showlang": 0,
    "startup": true,
    "startupfail": 1,
    "startupincorrect": 3,
    "startupshort": false,
    "startuptimeout": 10,
    "statusbar": false,
    "statusbarpos": "",
    "suppress": 0,
    "suppressblink": true,
    "suppressfocus": true,
    "suppresspopup": true,
    "suppresspopupremove": 30,
    "suppresssound": true,
    "suppresstemp": 10,
    "suppresstimer": 10,
    "toolsmenu": false,
    "unlockSync": false,
    "urlbar": true,
    "urlbarpos": "0go-button",
    "version": "1.29.2",
    "windows": {}
  },
  "Extension": [
    [
      "Enigmail",
      "1.9.8.3",
      "{847b3a00-7ab1-11d4-8f02-006008948af5}"
    ],
    [
      "Header Tools Lite",
      "0.6.1",
      "headerToolsLite{a}kaosmos.nnp"
    ],
    [
      "Lightning",
      "5.4.5",
      "{e2fda1a4-762b-4020-b5ad-a41df1933103}"
    ],
    [
      "Master Password+",
      "1.29.2",
      "masterpasswordtimeoutplus{a}vano"
    ]
  ],
  "Theme": [
    [
      "Default",
      "52.5.0",
      "{972ce4c6-7e08-4474-a285-3208198ce6fd}"
    ]
  ]
}
```

**Comment by vanowm** [comment]
Yes! Thank you very much, it's crashing now 💃

**Comment by vanowm** [comment]
Leaving this for latter today, but the issue seems to be not related to incorrect type of settings. Just activating show MP on startup with everything default still crashes TB. However enabling debug, stops crashing, no wonder I couldn't reproduce it before, had debug enabled all the time..

**Comment by etam** [comment]
A living example of heisenbug!

**Comment by vanowm** [comment]
I have some bad news, it seems the issue is not directly related to MP+ itself, but rather a bug in Thunderbird
I'm able crash it when showing prompt "too soon" on startup, using fresh profile with s simple extension that shows alert box on startup.

Also, on my computer if I wait 3 seconds (or more) after password prompt shown, it does NOT crash.

Good news is, I might have found a work around this bug, it doesn't have any side effects on my system, but I'd like you guys check it out on yours with v1.29b2. (But before you install it, make sure it still crashing for you when enabled startup prompt and you close prompt as soon as it appears, and it doesn't crash if you wait a few seconds before closing it)

**Comment by rotanid** [comment]
now i had the issue on another device with the exact same environment. by updating/downgrading a few times and deleting all settings and so on i managed to get it working, but i have no clue whatsoever which step brought the final fix :-(

**Comment by vanowm** [comment]
Can you confirm the issue still exists in [v1.29.3b2](https://github.com/vanowm/MasterPasswordPlus/commit/99fc4e780d961496091fa9d20e1321b619ebd9e8)?

**Comment by rotanid** [comment]
yes, i tried several versions, also your beta versions, then i removed everything, installed the newer version directly... i really don't know anymore in which order i did it, but finally it worked out :-(

**Comment by vanowm** [comment]
Do you have "Ask for password on startup" enabled? (extensions.masterPasswordPlus.startup) if not, can you test if with that option enabled it still works?

**Comment by rotanid** [comment]
that's my major option, so yes, it's enabled. other than this, i only set the "hide duplicates" thing

**Comment by wsmwk** [comment]
@vanowm Do you have a thunderbird bug# for crash ID for "the issue is not directly related to MP+ itself, but rather a bug in Thunderbird. I'm able crash it when showing prompt "too soon" on startup, using fresh profile with s simple extension that shows alert box on startup." ?

**Comment by rotanid** [comment]
it was working for me with your beta xpi, i just tried v1.30 but with this one i have a new problem: it doesn't ask for the masterpassword before starting, but asks after starting and doesn't accept the correct password, no matter how often i enter it.
i downgraded to your beta 1.29.3b2 xpi and it works again.

**Comment by vanowm** [comment]
@wsmwk no, not at the moment, can't find it...

@rotanid I've had similar issue when installed it on my main Waterfox profile. Fixed it by closing the browser and removing startupCache folder from it's profile.

**Comment by rotanid** [comment]
@vanowm there's no such folder. the error console showed, that the path to the preferences file is wrong. i checked it and the problem was, that the folder was called "Preferences" but the extension searched for "preferences" with lower case "p". i renamed it manually and now it works....

**Comment by vanowm** [comment]
That is very weird, preferences folder was always lower case.
Weirder yet, that previous version worked for you...I still think that since you manually updated extension, Thunderbird didn't clear properly extensions cache.
The startupCache folder on windows system might be somewhere in 
`%APPDATA%\Thunderbird`
or
`%LOCALAPPDATA%\Thunderbird`

**Comment by vanowm** [comment]
Indeed somehow default\preferences folder had capital letter...not sure how did that happen. Fixed in v1.30.1
